File: backend/src/main/java/com/horain/service/ProjectService.java
package com.horain.service;

import com.horain.dto.ProjectActivityTypeSummaryDto;
import com.horain.dto.ProjectDto;
import com.horain.model.Project;
import com.horain.repository.ProjectRepository;
import com.horain.repository.TimeLogRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.nio.ByteBuffer;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Service for project operations.
 */
@Service
public class ProjectService {

    private final ProjectRepository projectRepository;
    private final TimeLogRepository timeLogRepository;

    public ProjectService(ProjectRepository projectRepository, TimeLogRepository timeLogRepository) {
        this.projectRepository = projectRepository;
        this.timeLogRepository = timeLogRepository;
    }

    @Transactional
    public ProjectDto create(ProjectDto dto) {
        Project entity = new Project();
        entity.setName(dto.getName());
        entity.setDescription(dto.getDescription());
        entity.setBillable(dto.getBillable() != null ? dto.getBillable() : true);
        entity.setUserId(dto.getUserId());
        Instant now = Instant.now();
        entity.setCreatedAt(now);
        entity.setUpdatedAt(now);
        entity.setId(dto.getId() != null ? dto.getId() : UUID.randomUUID());
        Project saved = projectRepository.save(entity);
        return toDto(saved);
    }

    /** Idempotent create: skip if entity with same ID already exists. */
    @Transactional
    public void createOrSkip(String entityId, ProjectDto dto) {
        if (entityId == null || entityId.isBlank()) {
            create(dto);
            return;
        }
        UUID id = UUID.fromString(entityId);
        if (projectRepository.existsById(id)) return;
        dto.setId(id);
        create(dto);
    }

    @Transactional(readOnly = true)
    public Optional<ProjectDto> findById(UUID id) {
        return projectRepository.findById(id).map(this::toDto);
    }

    @Transactional(readOnly = true)
    public List<ProjectDto> findAll() {
        List<Project> all = projectRepository.findAll();
        Map<UUID, Long> revenueByProject = sumRevenueCentsByProjectMap();
        Map<UUID, Long> countByProject = countByProjectIdMap();
        Map<UUID, List<ProjectActivityTypeSummaryDto>> topActivityTypesByProject = topActivityTypesByProjectMap();
        return all.stream()
                .map(p -> toDto(p, revenueByProject.get(p.getId()), countByProject.get(p.getId()),
                        topActivityTypesByProject.get(p.getId())))
                .collect(Collectors.toList());
    }

    /** Builds projectId -> number of time log entries. Projects with zero logs are not in the result (default 0). */
    private Map<UUID, Long> countByProjectIdMap() {
        List<Object[]> rows = timeLogRepository.countByProjectId();
        return rows.stream()
                .collect(Collectors.toMap(
                        row -> toUuid(row[0]),
                        row -> ((Number) row[1]).longValue()));
    }

    private static final int TOP_ACTIVITY_TYPES_MAX = 5;

    /** Builds projectId -> list of top activity types (code, label, count) sorted by count desc, max 5. */
    private Map<UUID, List<ProjectActivityTypeSummaryDto>> topActivityTypesByProjectMap() {
        List<Object[]> rows = timeLogRepository.countByProjectIdAndActivityType();
        Map<UUID, List<ProjectActivityTypeSummaryDto>> byProject = new LinkedHashMap<>();
        for (Object[] row : rows) {
            UUID projectId = toUuid(row[0]);
            String code = row[1] != null ? row[1].toString() : "";
            String label = row[2] != null ? row[2].toString() : "";
            long count = row[3] != null ? ((Number) row[3]).longValue() : 0L;
            byProject.computeIfAbsent(projectId, k -> new ArrayList<>())
                    .add(new ProjectActivityTypeSummaryDto(code, label, count));
        }
        byProject.replaceAll((k, list) -> list.stream()
                .sorted(Comparator.comparingLong(ProjectActivityTypeSummaryDto::count).reversed())
                .limit(TOP_ACTIVITY_TYPES_MAX)
                .collect(Collectors.toList()));
        return byProject;
    }

    /** Builds projectId -> total revenue (cents) for billable entries with activity type. */
    private Map<UUID, Long> sumRevenueCentsByProjectMap() {
        List<Object[]> rows = timeLogRepository.sumRevenueCentsByProject();
        return rows.stream()
                .collect(Collectors.toMap(
                        row -> toUuid(row[0]),
                        row -> Math.round(((Number) row[1]).doubleValue())));
    }

    private static UUID toUuid(Object value) {
        if (value == null) throw new IllegalArgumentException("projectId is null");
        if (value instanceof UUID) return (UUID) value;
        if (value instanceof byte[] bytes && bytes.length == 16) {
            ByteBuffer bb = ByteBuffer.wrap(bytes);
            return new UUID(bb.getLong(), bb.getLong());
        }
        return UUID.fromString(value.toString());
    }

    /**
     * Updates an existing project. Only non-null fields in the patch are applied.
     */
    @Transactional
    public ProjectDto update(UUID id, ProjectDto patch) {
        Project entity = projectRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Project not found: " + id));
        if (patch.getName() != null && !patch.getName().isBlank()) {
            entity.setName(patch.getName().trim());
        }
        if (patch.getDescription() != null) {
            entity.setDescription(patch.getDescription());
        }
        if (patch.getBillable() != null) {
            entity.setBillable(patch.getBillable());
        }
        return toDto(projectRepository.save(entity));
    }

    /**
     * Deletes a project. Fails if the project has any time log entries (RESTRICT).
     */
    @Transactional
    public void deleteById(UUID id) {
        if (!projectRepository.existsById(id)) {
            throw new IllegalArgumentException("Project not found: " + id);
        }
        long count = timeLogRepository.countByProjectId(id);
        if (count > 0) {
            throw new IllegalStateException(
                    "Cannot delete project: it has " + count + " time log entries. Delete or reassign them first.");
        }
        projectRepository.deleteById(id);
    }

    /**
     * Fuzzy search by project name. Returns projects whose name contains the query (case-insensitive).
     */
    @Transactional(readOnly = true)
    public List<ProjectDto> searchByName(String name) {
        if (name == null || name.isBlank()) {
            return findAll();
        }
        return projectRepository.findByNameContainingIgnoreCase(name.trim()).stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    /**
     * When no exact/contains match exists, returns projects with similar names (typo-tolerant).
     * Uses normalized Levenshtein similarity; only returns projects above the similarity threshold.
     *
     * @param name       query (e.g. "Horian")
     * @param maxResults max number of close matches to return (e.g. 3)
     * @return list of projects sorted by similarity descending, best match first
     */
    @Transactional(readOnly = true)
    public List<ProjectDto> findCloseMatchesByName(String name, int maxResults) {
        if (name == null || name.isBlank() || maxResults <= 0) {
            return List.of();
        }
        String query = name.trim().toLowerCase();
        List<Project> all = projectRepository.findAll();
        double threshold = 0.5;
        return all.stream()
                .map(p -> new Object[]{p, similarity(query, p.getName().toLowerCase())})
                .filter(pair -> (Double) pair[1] >= threshold)
                .sorted(Comparator.<Object[], Double>comparing(pair -> (Double) pair[1]).reversed())
                .limit(maxResults)
                .map(pair -> toDto((Project) pair[0]))
                .toList();
    }

    /**
     * Normalized similarity between two strings (0 = unrelated, 1 = identical).
     * Based on Levenshtein distance: 1 - (distance / maxLength).
     */
    static double similarity(String a, String b) {
        if (a.isEmpty() && b.isEmpty()) return 1.0;
        if (a.isEmpty() || b.isEmpty()) return 0.0;
        int maxLen = Math.max(a.length(), b.length());
        int distance = levenshteinDistance(a, b);
        return 1.0 - (double) distance / maxLen;
    }

    private static int levenshteinDistance(CharSequence a, CharSequence b) {
        int n = a.length();
        int m = b.length();
        if (n == 0) return m;
        if (m == 0) return n;
        int[] prev = new int[m + 1];
        int[] curr = new int[m + 1];
        for (int j = 0; j <= m; j++) prev[j] = j;
        for (int i = 1; i <= n; i++) {
            curr[0] = i;
            for (int j = 1; j <= m; j++) {
                int cost = a.charAt(i - 1) == b.charAt(j - 1) ? 0 : 1;
                curr[j] = Math.min(Math.min(curr[j - 1] + 1, prev[j] + 1), prev[j - 1] + cost);
            }
            int[] swap = prev;
            prev = curr;
            curr = swap;
        }
        return prev[m];
    }

    private ProjectDto toDto(Project p) {
        return toDto(p, null, null, null);
    }

    private ProjectDto toDto(Project p, Long revenueCents, Long timeLogCount,
                             List<ProjectActivityTypeSummaryDto> topActivityTypes) {
        return ProjectDto.builder()
                .id(p.getId())
                .name(p.getName())
                .description(p.getDescription())
                .billable(p.getBillable() != null ? p.getBillable() : true)
                .createdAt(p.getCreatedAt())
                .updatedAt(p.getUpdatedAt())
                .userId(p.getUserId())
                .revenueCents(revenueCents)
                .timeLogCount(timeLogCount != null ? timeLogCount : 0L)
                .topActivityTypes(topActivityTypes != null ? topActivityTypes : List.of())
                .build();
    }
}

File: backend/src/main/kotlin/com/horain/service/ProjectService.kt
package com.horain.service

import com.horain.dto.ProjectActivityTypeSummaryDto
import com.horain.dto.ProjectDto
import com.horain.model.Project
import com.horain.repository.ProjectRepository
import com.horain.repository.TimeLogRepository
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.nio.ByteBuffer
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.Locale
import java.util.UUID
import kotlin.math.roundToLong

/**
 * Service for project operations.
 */
@Service
class ProjectService(
    private val projectRepository: ProjectRepository,
    private val timeLogRepository: TimeLogRepository
) {

    @Transactional
    fun create(dto: ProjectDto): ProjectDto {
        val entity = Project()
        entity.name = dto.name
        entity.description = dto.description
        entity.billable = dto.billable ?: true
        entity.userId = dto.userId
        val now = Instant.now()
        entity.createdAt = now
        entity.updatedAt = now
        entity.id = dto.id ?: UUID.randomUUID()
        val saved = projectRepository.save(entity)
        return toDto(saved)
    }

    /** Idempotent create: skip if entity with same ID already exists (or duplicate key on concurrent seed). */
    @Transactional
    fun createOrSkip(entityId: String?, dto: ProjectDto) {
        if (entityId.isNullOrBlank()) {
            create(dto)
            return
        }
        val id = UUID.fromString(entityId)
        try {
            if (projectRepository.existsById(id)) return
            dto.id = id
            create(dto)
        } catch (_: DataIntegrityViolationException) {
            // Already exists (e.g. concurrent seed, or flush during existsById revealed a duplicate); treat as skip
        }
    }

    @Transactional(readOnly = true)
    fun findById(id: UUID): java.util.Optional<ProjectDto> =
        projectRepository.findById(id).map { toDto(it) }

    @Transactional(readOnly = true)
    fun findAll(): List<ProjectDto> {
        val all = projectRepository.findAll()
        val revenueByProject = sumRevenueCentsByProjectMap()
        val countByProject = countByProjectIdMap()
        val topActivityTypesByProject = topActivityTypesByProjectMap()
        val durationByProject = sumDurationMinutesByProjectMap()
        return all.map { p ->
            toDto(
                p,
                revenueByProject[p.id],
                countByProject[p.id],
                topActivityTypesByProject[p.id],
                durationByProject[p.id]
            )
        }
    }

    /**
     * Projects that have at least one time log in [start, end), sorted by name (case-insensitive).
     * Revenue, counts, and top activity types are scoped to the same interval.
     */
    @Transactional(readOnly = true)
    fun findAllWithActivityInPeriod(start: Instant, end: Instant): List<ProjectDto> {
        require(start.isBefore(end)) { "activityFrom must be before activityTo" }
        require(ChronoUnit.DAYS.between(start, end) <= MAX_ACTIVITY_PERIOD_DAYS) {
            "Activity period must not exceed $MAX_ACTIVITY_PERIOD_DAYS days"
        }
        val projectIds = timeLogRepository.findDistinctProjectIdsByLoggedAtBetween(start, end).toSet()
        if (projectIds.isEmpty()) {
            return emptyList()
        }
        val projects = projectRepository.findAllById(projectIds)
            .sortedBy { it.name?.lowercase(Locale.ROOT) ?: "" }
        val revenueByProject = sumRevenueCentsByProjectMapForPeriod(start, end)
        val countByProject = countByProjectIdMapForPeriod(start, end)
        val topActivityTypesByProject = topActivityTypesByProjectMapForPeriod(start, end)
        val durationByProject = sumDurationMinutesByProjectMapForPeriod(start, end)
        return projects.map { p ->
            toDto(
                p,
                revenueByProject[p.id],
                countByProject[p.id],
                topActivityTypesByProject[p.id],
                durationByProject[p.id]
            )
        }
    }

    /** Builds projectId -> number of time log entries. Projects with zero logs are not in the result (default 0). */
    private fun countByProjectIdMap(): Map<UUID, Long> {
        val rows = timeLogRepository.countByProjectId()
        return rows.associate { row ->
            toUuid(row[0]) to (row[1] as Number).toLong()
        }
    }

    /** Builds projectId -> list of top activity types (code, label, count) sorted by count desc, max 5. */
    private fun topActivityTypesByProjectMap(): Map<UUID, List<ProjectActivityTypeSummaryDto>> {
        val rows = timeLogRepository.countByProjectIdAndActivityType()
        val byProject = linkedMapOf<UUID, MutableList<ProjectActivityTypeSummaryDto>>()
        for (row in rows) {
            val projectId = toUuid(row[0])
            val code = row[1]?.toString() ?: ""
            val label = row[2]?.toString() ?: ""
            val count = if (row[3] != null) (row[3] as Number).toLong() else 0L
            byProject.getOrPut(projectId) { mutableListOf() }
                .add(ProjectActivityTypeSummaryDto(code, label, count))
        }
        return byProject.mapValues { (_, list) ->
            list.sortedByDescending { it.count }.take(TOP_ACTIVITY_TYPES_MAX)
        }
    }

    /** Builds projectId -> total revenue (cents) for billable entries with activity type. */
    private fun sumRevenueCentsByProjectMap(): Map<UUID, Long> {
        val rows = timeLogRepository.sumRevenueCentsByProject()
        return rows.associate { row ->
            toUuid(row[0]) to (row[1] as Number).toDouble().roundToLong()
        }
    }

    private fun countByProjectIdMapForPeriod(start: Instant, end: Instant): Map<UUID, Long> {
        val rows = timeLogRepository.countLogsByProjectForPeriod(start, end)
        return rows.associate { row ->
            toUuid(row[0]) to (row[1] as Number).toLong()
        }
    }

    private fun topActivityTypesByProjectMapForPeriod(
        start: Instant,
        end: Instant
    ): Map<UUID, List<ProjectActivityTypeSummaryDto>> {
        val rows = timeLogRepository.countByProjectIdAndActivityTypeForPeriod(start, end)
        val byProject = linkedMapOf<UUID, MutableList<ProjectActivityTypeSummaryDto>>()
        for (row in rows) {
            val projectId = toUuid(row[0])
            val code = row[1]?.toString() ?: ""
            val label = row[2]?.toString() ?: ""
            val count = if (row[3] != null) (row[3] as Number).toLong() else 0L
            byProject.getOrPut(projectId) { mutableListOf() }
                .add(ProjectActivityTypeSummaryDto(code, label, count))
        }
        return byProject.mapValues { (_, list) ->
            list.sortedByDescending { it.count }.take(TOP_ACTIVITY_TYPES_MAX)
        }
    }

    private fun sumRevenueCentsByProjectMapForPeriod(start: Instant, end: Instant): Map<UUID, Long> {
        val rows = timeLogRepository.sumRevenueCentsByProjectForPeriod(start, end)
        return rows.associate { row ->
            toUuid(row[0]) to (row[1] as Number).toDouble().roundToLong()
        }
    }

    private fun sumDurationMinutesByProjectMap(): Map<UUID, Long> {
        val rows = timeLogRepository.sumDurationMinutesByProject()
        return rows.associate { row ->
            toUuid(row[0]) to (row[1] as Number).toLong()
        }
    }

    private fun sumDurationMinutesByProjectMapForPeriod(start: Instant, end: Instant): Map<UUID, Long> {
        val rows = timeLogRepository.sumDurationMinutesByProjectForPeriod(start, end)
        return rows.associate { row ->
            toUuid(row[0]) to (row[1] as Number).toLong()
        }
    }

    private fun toUuid(value: Any?): UUID {
        if (value == null) throw IllegalArgumentException("projectId is null")
        if (value is UUID) return value
        if (value is ByteArray && value.size == 16) {
            val bb = ByteBuffer.wrap(value)
            return UUID(bb.long, bb.long)
        }
        return UUID.fromString(value.toString())
    }

    /**
     * Updates an existing project. Only non-null fields in the patch are applied.
     */
    @Transactional
    fun update(id: UUID, patch: ProjectDto): ProjectDto {
        val entity = projectRepository.findById(id)
            .orElseThrow { IllegalArgumentException("Project not found: $id") }
        if (patch.name != null && patch.name!!.isNotBlank()) {
            entity.name = patch.name!!.trim()
        }
        if (patch.description != null) {
            entity.description = patch.description
        }
        if (patch.billable != null) {
            entity.billable = patch.billable
        }
        return toDto(projectRepository.save(entity))
    }

    /**
     * Deletes a project. Fails if the project has any time log entries (RESTRICT).
     */
    @Transactional
    fun deleteById(id: UUID) {
        if (!projectRepository.existsById(id)) {
            throw IllegalArgumentException("Project not found: $id")
        }
        val count = timeLogRepository.countByProjectId(id)
        if (count > 0) {
            throw IllegalStateException(
                "Cannot delete project: it has $count time log entries. Delete or reassign them first."
            )
        }
        projectRepository.deleteById(id)
    }

    /**
     * Fuzzy search by project name. Returns projects whose name contains the query (case-insensitive).
     */
    @Transactional(readOnly = true)
    fun searchByName(name: String?): List<ProjectDto> {
        if (name.isNullOrBlank()) {
            return findAll()
        }
        return projectRepository.findByNameContainingIgnoreCase(name.trim()).map { toDto(it) }
    }

    /**
     * When no exact/contains match exists, returns projects with similar names (typo-tolerant).
     * Uses normalized Levenshtein similarity; only returns projects above the similarity threshold.
     */
    @Transactional(readOnly = true)
    fun findCloseMatchesByName(name: String, maxResults: Int): List<ProjectDto> {
        if (name.isBlank() || maxResults <= 0) {
            return emptyList()
        }
        val query = name.trim().lowercase()
        val all = projectRepository.findAll()
        val threshold = 0.5
        return all.map { p -> p to similarity(query, p.name!!.lowercase()) }
            .filter { (_, sim) -> sim >= threshold }
            .sortedByDescending { it.second }
            .take(maxResults)
            .map { toDto(it.first) }
    }

    /**
     * Normalized similarity between two strings (0 = unrelated, 1 = identical).
     * Based on Levenshtein distance: 1 - (distance / maxLength).
     */
    fun similarity(a: String, b: String): Double {
        if (a.isEmpty() && b.isEmpty()) return 1.0
        if (a.isEmpty() || b.isEmpty()) return 0.0
        val maxLen = maxOf(a.length, b.length)
        val distance = levenshteinDistance(a, b)
        return 1.0 - distance.toDouble() / maxLen
    }

    private fun levenshteinDistance(a: CharSequence, b: CharSequence): Int {
        val n = a.length
        val m = b.length
        if (n == 0) return m
        if (m == 0) return n
        var prev = IntArray(m + 1) { j -> j }
        var curr = IntArray(m + 1)
        for (i in 1..n) {
            curr[0] = i
            for (j in 1..m) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                curr[j] = minOf(
                    curr[j - 1] + 1,
                    prev[j] + 1,
                    prev[j - 1] + cost
                )
            }
            val swap = prev
            prev = curr
            curr = swap
        }
        return prev[m]
    }

    private fun toDto(p: Project): ProjectDto = toDto(p, null, null, null, null)

    private fun toDto(
        p: Project,
        revenueCents: Long?,
        timeLogCount: Long?,
        topActivityTypes: List<ProjectActivityTypeSummaryDto>?,
        totalDurationMinutes: Long?
    ): ProjectDto =
        ProjectDto.builder()
            .id(p.id)
            .name(p.name)
            .description(p.description)
            .billable(p.billable ?: true)
            .createdAt(p.createdAt)
            .updatedAt(p.updatedAt)
            .userId(p.userId)
            .revenueCents(revenueCents)
            .timeLogCount(timeLogCount ?: 0L)
            .topActivityTypes(topActivityTypes ?: emptyList())
            .totalDurationMinutes(totalDurationMinutes ?: 0L)
            .build()

    companion object {
        private const val TOP_ACTIVITY_TYPES_MAX = 5
        const val MAX_ACTIVITY_PERIOD_DAYS = 366L
    }
}
